### ISSUE TITLE
Exception File "typing/ctype.ml", line 4735

### ISSUE THREAD

**Issue by EduardoRFS** [ocaml/ocaml#9774]
```sh
Fatal error: exception File "typing/ctype.ml", line 4735, characters 2-8: Assertion failed
```

Happens when trying to hide a type of a class

```
module A = struct class t = object method getWeight = 0.5 end end
module B = struct include A
                  type t end
```

**Comment by xavierleroy** [comment]
Paging a type system expert to triage this report.  Thank you.

**Comment by gasche** [comment]
@garrigue, @trefis or @lpw25 would be the experts here. I had a look nonetheless and will explain my findings below.

(Note that @lpw25 has a PR #8516 that does a big revamp of the class-signature representation and reportedly fixes several bugs, but has been waiting for a couple years now. If I understand correctly, progress there is stuck waiting for Leo to have more time again to look at it, and take early-review comments into account.)

Note: #9755 is basically a duplicate of this one with a slightly different reproduction case. I will close #9755, copying the reproduction case here (thanks!)

```ocaml
module type A  = sig class t : object method  getWeight : float end end
module type B  = sig include (A with type  t :=  < getWeight :float  >) end
```

Now what is happening:
- In the past, we would always error if a module had two declarations of the same name, except if they were both terms.
- In #1892 (merged in 4.08), @trefis implemented support for declaration B shadowing declaration A if A comes from an `include`. This calls the `nondep_*` from the module-system logic, to remove all references to A in the remaining signature.
- It seems that this mode of use of was not planned by the `nondep_*` for classes or classes type: they simply `assert` that the name being removed is not their name (but a type that may be referenced in their own type, etc.). This assertion now fails.

Two further remarks:
- the "assert" for classes and class types was placed by Jérôme Vouillon in 1998 ( 87b17301f4a408109a0f00f7a297b02c79724bc1 )
- because classes (not just class types) are also affected, another repro case is

```ocaml
module A = struct class t = objet end end
module B = struct include A type t end
```

Naively, I would propose to fix the issue by removing the `assert`, and doing the right thing here (which I think is raising an exception, which would be caught by the type-checker later on and result in the declaration A being dropped as it should be (so the final signature would be just `sig type te end`. Of course, I have no idea whether some parts of the type-checker really rely on this assumption and could go horribly wrong if we remove it. Exciting! I will try to propose a PR later on.

**Comment by trefis** [comment]
I just tried to reproduce the issue without using an include and shadowing, I failed:
```
# module type A = sig class t : object end end;;
module type A = sig class t : object  end end
# module A = struct class t = object end end;;
module A : sig class t : object  end end
# module B = (A : A with type t := A.t);;
Fatal error: exception Invalid_argument("split_at")
Raised at Misc.Stdlib.List.split_at.aux in file "utils/misc.ml", line 147, characters 18-53
Called from Printtyp.skip in file "typing/printtyp.ml", line 1626, characters 19-50
Called from Printtyp.protect_rec_items.get_ids in file "typing/printtyp.ml", line 1632, characters 66-78
Called from Printtyp.protect_rec_items in file "typing/printtyp.ml", line 1634, characters 41-67
Called from Printtyp.tree_of_signature_rec in file "typing/printtyp.ml", line 1681, characters 6-29
Called from Printtyp.wrap_env in file "typing/printtyp.ml", line 1571, characters 13-22
Called from Printtyp.tree_of_modtype in file "typing/printtyp.ml", line 1652, characters 27-47
Called from Printtyp.tree_of_module in file "typing/printtyp.ml", line 1719, characters 30-59
Called from Printtyp.trees_of_sigitem in file "typing/printtyp.ml", line 1702, characters 7-48
Called from Printtyp.print_items.print in file "typing/printtyp.ml", line 1753, characters 18-39
Called from Toploop.execute_phrase.(fun) in file "toplevel/toploop.ml", line 307, characters 40-60
Called from Misc.try_finally in file "utils/misc.ml", line 31, characters 8-15
Re-raised at Misc.try_finally in file "utils/misc.ml", line 45, characters 10-56
Called from Toploop.execute_phrase in file "toplevel/toploop.ml", line 289, characters 16-842
Re-raised at Toploop.execute_phrase in file "toplevel/toploop.ml", line 332, characters 32-39
Called from Toploop.execute_phrase in file "toplevel/toploop.ml", line 374, characters 6-42
Re-raised at Toploop.execute_phrase in file "toplevel/toploop.ml", line 377, characters 4-13
Called from Toploop.loop in file "toplevel/toploop.ml", line 615, characters 12-41
Re-raised at Location.report_exception.loop in file "parsing/location.ml", line 926, characters 14-25
Called from Toploop.loop in file "toplevel/toploop.ml", line 620, characters 11-42
Called from Topstart in file "toplevel/topstart.ml", line 16, characters 8-22
```
(This looks like a completely different issue, but I found it funny)

**Comment by trefis** [comment]
Also, note that if I change the definition of `B` in my previous example to:
```ocaml
module B = struct include (A : A with type t := A.t) type t end
```
then I get:
```ocaml
module B : sig class t : object  end end
```

I *guess* what happens is that the class gets removed, but not the class type, nor the `t#` version of it.
Then in this last example, we reintroduce a type `t`, so the printer does its silly magic: « oh, a t, a class type t, a t#, let me just print  "class t" », whereas in the previous example it got all confused and crashed?

This is really just guess work.

**Comment by gasche** [comment]
Well if you want to learn more about Printtyp, feel free to lend a hand to the semi-big PR there, #8929 :-)

My take-away from your examples is that your `include` change is not the only language change that resulted in `nondep_*` being used more for classes (than in 1998), and that there are other weird side-effects, not just this `assert`. I still plan to work on removing the `assert` and doing the "proper" thing in this case, which should fix at least a part of the issues.

**Comment by trefis** [comment]
To illustrate what I was saying earlier, and to prompt you to define what the "proper" thing should be for classes, let me put that here:
```
# module type A = sig class t : object end end;;
module type A = sig class t : object  end end
# module A = struct class t = object end end;;
module A : sig class t : object  end end
# let module B = (A : A with type t := A.t) in (object end : B.t);;
Error: Unbound type constructor B.t
# let module B = (A : A with type t := A.t) in (object end : #B.t);;
- : A.t = <obj>
# let module B = (A : A with type t := A.t) in object inherit B.t end;;
- : A.t = <obj>
```

I'm not saying this is the right behavior, I'm just saying the issue was swept under the rug; and that the assertion doesn't trigger in this case because only the `t` ident corresponding to the type gets removed, not the other ones. I'm also *not* saying the the shadowing done on included items is doing the right thing either, it could also be hiding only one `t`, just a different one.

**Comment by gasche** [comment]
In my mind all the "shadow" objects that come from a given "real" definition should be kept only if we keep the real definition (and conversely, the definition should not be kept if one of the shadow objects cannot be erased from the signature). I don't know if it is possible/pleasant to implement this behavior with the current representation of the shadow objects.

**Comment by trefis** [comment]
I agree with you, and I don't know either (I haven't looked at that code in a long while).
I volunteer to review if you clean up that whole mess.

**Comment by gasche** [comment]
I looked at this more this afternoon.

I think that the right place to fix the issue would be in the `Signature_names` module of `Typemod`, which iterates on a signature and accumulates state to decide which of the signature items are shadowed and should be "removed" (or error we don't want to allow shadowing). This code should know about "ghost" declarations for classes and class types, and deal with them as a group: if one of them is marked for removal, they should all be marked.

Note: "ghost" declarations are generated when translating the Parsetree in a Typedtree. The `Typeclass` module is the one creating the names for the ghost items, and for each class (class type) declaration it returns metadata with the names of the ghost items. But then the information that they are "ghost" in lost in the Typedtree proper: they are not marked as ghost, handled as any other item, and there is no place that stores the mapping from the class (class type) declaration to the name of its ghost items. (There are ugly hacks in the type system to work around this lack of metadata, see for example [Env.find_hash_type](https://github.com/ocaml/ocaml/blob/4822a88/typing/env.ml#L1060-L1074).)

I considered the following approaches:
- Attempt 1: change the callers of `Signature_names.check` to use a `check_group` function on the whole group, and then the module internally knows about groups and does the right thing. This does not work: while the typing code for `Psig_class` will generates the whole group, and could check it altogether, the typing code for `Psig_include` extracts a *typed* signature and iterates on all its item, so it does not know about the ghost/group structure.
- Attempt 2: change `Signature.check_class{,_type}` to do extra work to also register the group. This does not work because the typedtree does not store any information to recover the ghost item identifiers from the class (class type) item.
- Attempt 3: store the ghost-item-names metadata in class (class type) declarations in the typedtree. I don't know how to extend substitution/renaming/freshening to properly deal with this metadata. At the point where you see the class declaration, the ghost items have not been traversed yet; if we want to "freshen" them, we would have to pick the fresh name right now, and then pass this information along to the ghost items in the remainder of the list to be processed. This adds a new case to the handling of all syntactic categories that may be ghost item, which sounds invasive and easy to get wrong. This may be the right approach, but I'm not impatient to go there. Plus there are ugly warts lying around in the code that I haven't tried understand in details, and will make this more painful than it should be: see #6650 for example. Abort.
- Attempt 4: use the ordering of signature elements to recreate the group structure, by simply assuming that the 4 (3) items coming after `Sig_class` (`Sig_class_type`) are its corresponding ghost items (we can do partial dynamic checks of this assumption). Use this group to check typed signatures. This is how `Printtyp` deals with the same problem, see the function [Printtyp.filter_rem_sig](https://github.com/ocaml/ocaml/blob/4822a88/typing/printtyp.ml#L1625-L1632). I thought that it would be nice to share code for this logic: it relies on hackish implicit invariants, so I would rather not have two different implementations of it in the codebase. But then I realized that @Octachron's cleanup PR on Printtyp #8929 changes this function completely, and that the new implementation seems nicer but has some details that I don't completely understand (see [this review question](https://github.com/ocaml/ocaml/pull/8929#discussion_r495428656)). I believe that it is wiser to wait for #8929 to settle down before we try extracting the group-signature-items logic for reuse in other parts of the type-checker. Yield.

My current understanding is that the best course of action is to warmly encourage @Octachron to go back to #8929 before I work on the present issue again.

**Comment by lpw25** [comment]
Probably you've already discovered this behaviour, but there is a related issue with destructive substitution: #6654.

In the medium term, I think it would be good to change types.ml to represent these grouped items directly as a single item. That should force us to go around and fix all the places where we might try to remove only part of them. (A similar thing for the #row types from private row types would also be good for the same reasons.).

For now, doing some small fixes like the one outlined by Gabriel is probably the way to go.

(Re. PR #8516. I still fully intend to get back to this one. Just struggling to find the time for it...)

**Comment by gasche** [comment]
@Octachron now that your work on adding more structure to signature items has been merged ( #10219 ), it may be worth revisiting the present issue. Is it something that you are interested on? Should we try to find some time to discuss it together? (I don't presently have time, so I'm not pushing for this option in particular.)

I would be tempted to try to tackle the present issue by fixing the implementation of `Includemod.Signature_names`, by using your grouping code from #10219 (this may require moving it to a more central place), and see if that works. Once we have experience with this approach, we could consider using it in other situations (basically, most traversals of signature items could benefit), or whether we want to change the signature representation by storing the groups directly instead of re-creating them. (The representation change sounds like the best long-term choice, but I think it would be more palatable and incremental to not go for it first, but instead try the approach on one specific issue.)

**Comment by Octachron** [comment]
@gasche , I was thinking of  having a look at this issue indeed. If you want to discuss it at some point, you are welcome.

**Comment by gasche** [comment]
Github automatically closed this issue on account of @Octachron's commit message in #10385 containing both the words "fixes" and "#9974" *somewhere*, which sounds a bit light. The issue reported and variants in the thread are still present.

**Comment by Octachron** [comment]
Closed by #10401 .
